test_launch_ros/test/test_launch_ros/actions/test_ros_timer_action.py

```diff
@@ -0,0 +1,201 @@
+            period=1.,
+            actions=[
+                launch.actions.Shutdown(reason='fast shutdown')
+            ]
+        ),
+
+        RosTimerAction(
+            period=2.,
+            actions=[
+                launch.actions.Shutdown(reason='slow shutdown')
+            ]
+        ),
+
+        _shutdown_listener_factory(shutdown_reasons),
+    ])
+
+    ls = launch.LaunchService()
+    ls.include_launch_description(ld)
+    assert 0 == ls.run()
+    assert len(shutdown_reasons) == 1
+    assert shutdown_reasons[0].reason == 'fast shutdown'
+
+
+def test_time_is_passing():
+    """Test that time is actually passing (sanity check)."""
+    # This test starts a 5 second timer and verifies that
+    # at least 5 seconds have passed between start of test
+    # and shutdown
+    shutdown_reasons = []
+
+    ld = launch.LaunchDescription([
+        launch.actions.ExecuteProcess(
+            cmd=[sys.executable, '-c', 'while True: pass'],
+        ),
+
+        RosTimerAction(
+            period=5.,
+            actions=[
+                launch.actions.Shutdown(reason='Five second timeout')
+            ]
+        ),
+
+        _shutdown_listener_factory(shutdown_reasons),
+    ])
+
+    start_time = time.time()
+    ls = launch.LaunchService()
+    ls.include_launch_description(ld)
+    assert 0 == ls.run()
+    assert abs(time.time() - start_time) > 5
+    assert shutdown_reasons[0].reason == 'Five second timeout'
+
+
+def test_timer_uses_sim_time():
+    """Test that timer uses time from /clock topic."""
+    # Create clock publisher node
+    rclpy.init()
+    node = rclpy.create_node('clock_publisher_node')
+    publisher = node.create_publisher(ClockMsg, '/clock', 10)
+
+    # Increment sim time by 100 every time callback is called
+    def timer_callback(publisher, time_msg):
+        time_msg.sec += 100
+        publisher.publish(ClockMsg(clock=time_msg))
+
+    # For every second of system time, publish new sim time value
+    callback_clock = Clock(clock_type=ClockType.SYSTEM_TIME)
+    time_msg = Time(sec=0, nanosec=0)
+    node.create_timer(1, partial(timer_callback, publisher, time_msg), clock=callback_clock)
+
+    thread = threading.Thread(target=rclpy.spin, args=(node, ), daemon=True)
+    thread.start()
+
+    ld = launch.LaunchDescription([
+
+        launch.actions.ExecuteProcess(
+            cmd=[sys.executable, '-c', 'while True: pass'],
+        ),
+
+        RosTimerAction(
+            period=200.0,
+            actions=[
+                launch.actions.Shutdown(reason='timer expired')
+            ],
+            use_sim_time=True  # Must be set to allow timer action to use sim time
+        ),
+    ])
+
+    start_time = time.time()
+
+    ls = launch.LaunchService()
+    ls.include_launch_description(ld)
+    assert 0 == ls.run()
+
+    # Simulated time is 100x faster, so 200 sec timer should finish in 2 sec
+    # (Some extra time provided)
+    assert (time.time() - start_time < 6)
```

Comment:
6 second seems like a very large tolerance, is that necessary?

Reply:
I changed these tests so that the start and end times are now recorded using the `OpaqueFunction` action while the launch service is running. Measuring the timeout is now much more accurate, so the tolerance can be smaller.